File: runtime/kotlin/src/main/java/org/apache/camel/k/kotlin/KotlinRoutesLoader.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.camel.k.kotlin;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.Collections;
import java.util.List;
import javax.script.Bindings;
import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import javax.script.SimpleBindings;

import org.apache.camel.CamelContext;
import org.apache.camel.builder.RouteBuilder;
import org.apache.camel.k.jvm.dsl.Components;
import org.apache.camel.k.jvm.Language;
import org.apache.camel.k.jvm.RoutesLoader;
import org.apache.camel.util.ResourceHelper;

public class KotlinRoutesLoader implements RoutesLoader {

    @Override
    public List<Language> getSupportedLanguages() {
        return Collections.singletonList(Language.Kotlin);
    }

    @Override
    public RouteBuilder load(String resource) throws Exception {
        return new RouteBuilder() {
            @Override
            public void configure() throws Exception {
                final CamelContext context = getContext();
                final ScriptEngineManager manager = new ScriptEngineManager();
                final ScriptEngine engine = manager.getEngineByExtension("kts");
                final Bindings bindings = new SimpleBindings();

                // Exposed to the underlying script, but maybe better to have
                // a nice dsl
                bindings.put("builder", this);
                bindings.put("context", context);
                bindings.put("components", new Components(context));

                try (InputStream is = ResourceHelper.resolveMandatoryResourceAsInputStream(context, resource)) {
                    StringBuilder builder = new StringBuilder();

                    // extract global objects from 'bindings'
                    builder.append("val builder = bindings[\"builder\"] as org.apache.camel.builder.RouteBuilder").append('\n');
                    builder.append("val context = bindings[\"context\"] as org.apache.camel.CamelContext").append('\n');
                    builder.append("val components = bindings[\"components\"] as org.apache.camel.k.jvm.dsl.Components").append('\n');

                    // create aliases for common functions
                    builder.append("fun from(uri: String): org.apache.camel.model.RouteDefinition = builder.from(uri)").append('\n');
                    builder.append("fun rest(): org.apache.camel.model.rest.RestDefinition = builder.rest()").append('\n');
                    builder.append("fun restConfiguration(): org.apache.camel.model.rest.RestConfigurationDefinition = builder.restConfiguration()").append('\n');

                    engine.eval(builder.toString(), bindings);
                    engine.eval(new InputStreamReader(is), bindings);
                }
            }
        };
    }
}

File: runtime/kotlin/src/main/kotlin/org/apache/camel/k/kotlin/KotlinRoutesLoader.kt
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.camel.k.kotlin

import org.apache.camel.builder.RouteBuilder
import org.apache.camel.k.jvm.Language
import org.apache.camel.k.jvm.RoutesLoader
import org.apache.camel.k.jvm.RuntimeRegistry
import org.apache.camel.k.kotlin.dsl.IntegrationConfiguration
import org.apache.camel.util.ResourceHelper
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.File
import java.io.InputStreamReader
import kotlin.script.experimental.api.*
import kotlin.script.experimental.host.toScriptSource
import kotlin.script.experimental.jvm.dependenciesFromClassloader
import kotlin.script.experimental.jvm.javaHome
import kotlin.script.experimental.jvm.jvm
import kotlin.script.experimental.jvmhost.BasicJvmScriptEvaluator
import kotlin.script.experimental.jvmhost.BasicJvmScriptingHost
import kotlin.script.experimental.jvmhost.JvmScriptCompiler

class KotlinRoutesLoader : RoutesLoader {
    companion object {
        val LOGGER : Logger = LoggerFactory.getLogger(KotlinRoutesLoader::class.java)
    }

    override fun getSupportedLanguages(): List<Language> {
        return listOf(Language.Kotlin)
    }

    @Throws(Exception::class)
    override fun load(registry: RuntimeRegistry, resource: String): RouteBuilder {
        return object : RouteBuilder() {
            @Throws(Exception::class)
            override fun configure() {
                val builder = this
                val compiler = JvmScriptCompiler()
                val evaluator = BasicJvmScriptEvaluator()
                val host = BasicJvmScriptingHost(compiler = compiler, evaluator = evaluator)
                val javaHome = System.getenv("KOTLIN_JDK_HOME") ?: "/usr/lib/jvm/java"

                LOGGER.info("JAVA_HOME is set to {}", javaHome)

                ResourceHelper.resolveMandatoryResourceAsInputStream(context, resource).use { `is` ->
                    val result = host.eval(
                        InputStreamReader(`is`).readText().toScriptSource(),
                        ScriptCompilationConfiguration {
                            baseClass(IntegrationConfiguration::class)
                            jvm {
                                //
                                // This is needed as workaround for:
                                //     https://youtrack.jetbrains.com/issue/KT-27497
                                //
                                javaHome(File(javaHome))

                                //
                                // The Kotlin script compiler does not inherit
                                // the classpath by default
                                //
                                dependenciesFromClassloader(wholeClasspath = true)
                            }
                        },
                        ScriptEvaluationConfiguration {
                            //
                            // Arguments used to initialize the script base class
                            //
                            constructorArgs(registry, builder)
                        }
                    )

                    for (report in result.reports) {
                        if (report.severity == ScriptDiagnostic.Severity.ERROR) {
                            LOGGER.error("{}", report.message, report.exception)
                        } else if (report.severity == ScriptDiagnostic.Severity.WARNING) {
                            LOGGER.warn("{}", report.message, report.exception)
                        } else {
                            LOGGER.info("{}", report.message)
                        }
                    }
                }
            }
        }
    }
}
